src/charm.py

```diff
@@ -121,24 +70,116 @@ def _on_update_status(self, event):
+        config = self.model.config
+        peers_data = self.model.get_relation("mysql").data[self.app]
+        env_config = {}
+        env_config["MYSQL_ROOT_PASSWORD"] = peers_data["mysql_root_password"]
+
+        if config.get("MYSQL_USER") and config.get("MYSQL_PASSWORD"):
+            env_config["MYSQL_USER"] = config["MYSQL_USER"]
+            env_config["MYSQL_PASSWORD"] = config["MYSQL_PASSWORD"]
+
+        if config.get("MYSQL_DATABASE"):
+            env_config["MYSQL_DATABASE"] = config["MYSQL_DATABASE"]
+
+        return env_config
+
+    ##############################################
+    #             UTILITY METHODS                #
+    ##############################################
+    def _mysql_root_password(self) -> str:
+        """
+        Returns MYSQL_ROOT_PASSWORD from the config,
+        if the password isn't in StoredState, generates one.
+        """
+        password_from_config = self.config["MYSQL_ROOT_PASSWORD"]
+        if password_from_config:
+            logger.debug("Retriving MYSQL_ROOT_PASSWORD from config")
+            return password_from_config
+        else:
+            logger.debug("MYSQL_ROOT_PASSWORD generated")
+            return MySQL.new_password(20)
+
+    def _update_peers(self):
+        if self.unit.is_leader():
+            peers_data = self.model.get_relation("mysql").data[self.app]
+
+            if not peers_data.get("mysql_root_password"):
+                peers_data["mysql_root_password"] = self._mysql_root_password()
+
+    def _configure_pod(self):
+        """Configure the Pebble layer for MySQL."""
+        if not self._stored.pebble_ready:
+            self.unit.status = MaintenanceStatus(
+                "Waiting for Pod startup to complete"
+            )
+            return False
+
+        layer = self._build_pebble_layer()
+
+        if not layer["services"]["mysql"]["environment"].get(
+            "MYSQL_ROOT_PASSWORD", False
+        ):
+            self.unit.status = MaintenanceStatus(
+                "Awaiting leader node to set MYSQL_ROOT_PASSWORD"
+            )
+            return False
+
+    def _build_pebble_layer(self):
+        """Construct the pebble layer"""
+        logger.debug("Building pebble layer")
+        return {
+            "summary": "MySQL layer",
+            "description": "Pebble layer configuration for MySQL",
+            "services": {
+                "mysql": {
+                    "override": "replace",
+                    "summary": "mysql service",
+                    "command": "docker-entrypoint.sh mysqld",
+                    "startup": "enabled",
+                    "environment": self.env_config,
+                },
+            },
+        }
+
     def _provide_mysql(self) -> None:
-        if self._stored.mysql_initialized:
+        if self._is_mysql_initialized():
             self.mysql_provider = MySQLProvider(
                 self, "database", "mysql", self.mysql.version()
             )
```

Comment:
I would avoid partial construction if possible.
A related warning that PyCharm has in this case is:
"Instance attribute variable defined outside __init__ "